cmd/consul-dataplane/config.go

```diff
@@ -0,0 +1,294 @@
+	DefaultGracefulPort                       = 20300
+	DefaultDumpEnvoyConfigOnExitEnabled       = false
+
+	DefaultUseCentralTelemetryConfig = true
+	DefaultPromRetentionTime         = 60 * time.Second
+	DefaultPromScrapePath            = "/metrics"
+	DefaultPromMergePort             = 20100
+)
+
+// buildDataplaneConfig builds the necessary config needed for the
+// dataplane to start. We begin with the default version of the dataplane
+// config(with the default values) followed by merging the file based
+// config generated from the `-config-file` input into it.
+// Since values given via CLI flags take the most precedence, we finally
+// merge the config generated from the flags into the previously
+// generated/merged config
+func (f *FlagOpts) buildDataplaneConfig() (*consuldp.Config, error) {
+	var consuldpConfig, consuldpCfgFromFlags *consuldp.Config
+
+	consuldpConfig = buildDefaultConsulDPConfig()
+	consuldpCfgFromFlags = f.buildConfig()
+
+	if f.configFile != "" {
+		consuldpCfgFromFile, err := f.buildConfigFromFile()
+		if err != nil {
+			return nil, err
+		}
+
+		mergeConfigs(consuldpConfig, consuldpCfgFromFile)
+	}
+
+	mergeConfigs(consuldpConfig, consuldpCfgFromFlags)
+
+	return consuldpConfig, nil
+}
+
+// Constructs a config based on the values present in the config json file
+func (f *FlagOpts) buildConfigFromFile() (*consuldp.Config, error) {
+	var cfg *consuldp.Config
+	data, err := os.ReadFile(f.configFile)
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(data, &cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// Constructs a config based on the values given via the CLI flags
+func (f *FlagOpts) buildConfig() *consuldp.Config {
+	return &consuldp.Config{
+		Consul: &consuldp.ConsulConfig{
+			Addresses: f.addresses,
+			GRPCPort:  f.grpcPort,
+			Credentials: &consuldp.CredentialsConfig{
+				Type: consuldp.CredentialsType(f.credentialType),
+				Static: consuldp.StaticCredentialsConfig{
+					Token: f.token,
+				},
+				Login: consuldp.LoginCredentialsConfig{
+					AuthMethod:      f.loginAuthMethod,
+					Namespace:       f.loginNamespace,
+					Partition:       f.loginPartition,
+					Datacenter:      f.loginDatacenter,
+					BearerToken:     f.loginBearerToken,
+					BearerTokenPath: f.loginBearerTokenPath,
+					Meta:            f.loginMeta,
+				},
+			},
+			ServerWatchDisabled: f.serverWatchDisabled,
+			TLS: &consuldp.TLSConfig{
+				Disabled:           f.tlsDisabled,
+				CACertsPath:        f.tlsCACertsPath,
+				ServerName:         f.tlsServerName,
+				CertFile:           f.tlsCertFile,
+				KeyFile:            f.tlsKeyFile,
+				InsecureSkipVerify: f.tlsInsecureSkipVerify,
+			},
+		},
+		Service: &consuldp.ServiceConfig{
+			NodeName:  f.nodeName,
+			NodeID:    f.nodeID,
+			ServiceID: f.serviceID,
+			Namespace: f.namespace,
+			Partition: f.partition,
+		},
+		Logging: &consuldp.LoggingConfig{
+			Name:     "consul-dataplane",
```

Comment:
This string "consul-dataplane" is at 5 places, we should move this to a constant.